Direct to firrtl

jackkoenig opened the pull request:
I think this still needs a little work, in particular the function converting `Command`s to `Statement`s is a bit of a beast. I would appreciate suggestions on how to improve it! I'll provide some benchmarking numbers soon so that we can ensure this doesn't hurt performance (although for larger designs, skipping the parser should only help).

I should note that the "Converter" as currently implemented is kind of a transliteration of the existing Emitter. I would greatly appreciate thoughts on improvements. It is slightly more verbose than the Emitter because some emission logic lives within the Chisel IR and at least initially I decided to keep all conversion logic contained.

<!-- choose one -->
**Type of change**: other enhancement

<!-- choose one -->
**Impact**: API addition (no impact on existing code) -- You can now get FIRRTL IR directly in Chisel!

<!-- choose one -->
**Development Phase**: implementation

**Release Notes**
Directly translate from ChiselIR to FIRRTL without using the FIRRTL ANTLR Parser

chick commented on src/main/scala/chisel3/internal/firrtl/Converter.scala:
```diff
@@ -0,0 +1,238 @@
+  // TODO refactor
+  def convert(cmds: Seq[Command], ctx: Component): fir.Statement = {
+    @tailrec
+    def rec(acc: Queue[fir.Statement],
```
What does `rec` mean?

chick replied:
You've probably considered this but could you create a trait like HasToStatement that most of the Commands could extend. Then the convert of command to statement could handle most of the cases with 
```
  case m: HasToStatement =>
     rec(acc :+ m.toStatement, scope)(cmds.tail)
```
It changes the encapsulation of information, since for example, DefWire has to know how to make a fir.DefWire but it sure cleans up the massive match. I can see it's a bit ugly though since toStatement would need access to the Converter object for info, clocks etc.

jackkoenig replied:
Yeah I have thought about it a bit but your point is well taken and I ought to share some of my thinking.

In prototyping this, I wanted to touch as little code as possible which is why the Converter is wholly self-contained, I am open to some amount of encapsulation though. `toStatement` would have to take the `Component` as an argument, but that's how `def fullName` currently works for converting `Args` to `Expressions`.

A big question related to performance I have that goes back to `serialize` in FIRRTL is whether large match statements or inheritance are better for performance. Serialization functions in this case are almost certainly megamorphic which is generally terrible for performance, but when we moved serialize from a match statement to a method on `FIRRTLNode`, it still got faster. Perhaps the dynamic method lookup in megamorphic state is still faster in practice than large match blocks. I'm really not sure.

jackkoenig commented:
Currently the performance of this is terrible because the `Seq` of `Commands` is an `ArrayBuffer`, so repeatedly calling `.tail` on it is super bad. Simple solution is to convert to `List` before recursing, considering other options as well.

ducky64 commented on src/main/scala/chisel3/internal/firrtl/Converter.scala:
```diff
@@ -0,0 +1,238 @@
+
+  // Whens markers are flat so scope must be inferred
+  // TODO refactor
+  def convert(cmds: Seq[Command], ctx: Component): fir.Statement = {
```
There's quite a bit of code here, can you write a summary of the algorithm in comments?
This mostly seems to be a straightforward 1:1 map of Chisel to FIRRTL types except with some funny stuff for whens.

jackkoenig replied:
Yeah this definitely needs a summary, will add

ducky64 commented on src/main/scala/chisel3/internal/firrtl/Converter.scala:
```diff
@@ -0,0 +1,238 @@
+          val con = fir.Connect(convert(info), convert(loc, ctx), convert(exp, ctx))
+          rec(acc :+ con, scope)(cmds.tail)
+        case BulkConnect(info, loc, exp) =>
+          val con = fir.PartialConnect(convert(info), convert(loc, ctx), convert(exp, ctx))
```
Are FIRRTL PartialConnect semantics equivalent to Chisel BulkConnect semantics? Or should this exception out because this should never be seen?

jackkoenig replied:
FIRRTL PartialConnect semantics are closer to Chisel2 (and the current `Chisel._` `<>`) semantics actually. It's still used by the compatibility wrapper.

jackkoenig commented:
I guess I should write down some of my thinking on this and let y'all know where I'm going with this:

It appears that the current ANTLR Parser for FIRRTL is the main culprit in the high memory use for FIRRTL. There are two (obvious) possible solutions:

1. Reduce the parser memory use
1. Don't use the parser

I spent some time trying to improve the parser (see https://github.com/freechipsproject/firrtl/pull/824), but due to past experience and pain with the parser, I've been working on number 2. For this, I saw two main ways to do it:

1. Convert from Chisel to FIRRTL in memory and invoke FIRRTL in the same JVM
1. Use ProtoBuf (or other library) for serialization and deserialization

I originally was working on number 1 (see this PR), but it turns out to be a pretty complex game of wack-a-mole with static references to Chisel objects that prevent garbage collection (harming memory use substantially). To illustrate, using a large design, I get the following memory stats:

* Elaborated Chisel Objects: 16 GB
* FIRRTL Representation: 3.5 GB

When I invoke them in the same JVM, the elaborated Chisel graph sticks around the whole time. It's probably possible to fix them all, but due to this issue as well as the general notion that it is simpler on the build system to keep Chisel and FIRRTL split, I decided to refocus on number 2. I am in the process of providing a translator from FIRRTL to ProtoBuf. The plan is to just implement FIRRTL<->Protobuf and use this PR to translate Chisel to FIRRTL for serialization.

For the record, the Chisel object is still massive but I suspect there are easy gains there. Even the FIRRTL representation has plenty of room for improvement.

jackkoenig commented:
retest this please

jackkoenig commented on src/main/scala/chisel3/internal/firrtl/Converter.scala:
```diff
@@ -0,0 +1,238 @@
+  def convert(bp: BinaryPoint): fir.Width = bp match {
+    case UnknownBinaryPoint => fir.UnknownWidth
+    case KnownBinaryPoint(value) => fir.IntWidth(value)
+  }
```
Yeah that was my thinking, a later refactor can clean a lot of stuff related to this and the Emitter

jackkoenig commented on src/main/scala/chisel3/internal/firrtl/Converter.scala:
```diff
@@ -0,0 +1,238 @@
+      val lit = bp.asInstanceOf[KnownBinaryPoint].value
+      fir.DoPrim(firrtl.PrimOps.AsFixedPoint, Seq(uint), Seq(lit), fir.UnknownType)
+    case lit: ILit => throwException(s"Internal Error! Unexpected ILit: $lit")
+  }
```
Yeah I agree the density is pretty tough. I ended up deciding to have all of the cases be multi-line. The matches were so different in length that alignment on `=>` wasn't working very well

jackkoenig commented:
I've refactored the style of the convert method for Commands to Statements (since that is by far the densest part). I split out conversion for Commands that map 1:1 to Statements from the when scope handling logic which is much more complicated. I also added clarifying comments to describe the basic process and noted things that I feel are non-obvious (or inconsistent). I also added `Driver.dumpProto` (which is why this is currently failing, it will pass once https://github.com/freechipsproject/firrtl/pull/832 is merged).

I have also posted performance numbers in https://github.com/freechipsproject/firrtl/pull/832. I backed out of using FIRRTL serialize because it is like 4x slower than the Chisel Emitter (as noted in https://github.com/freechipsproject/firrtl/issues/838). However, the converter is still used by `Driver.execute` and `Driver.dumpProto`. I think speeding up FIRRTL serialization so that the Chisel Emitter can be removed is desirable but beyond the scope of this PR

jackkoenig commented:
This is ready to go
